### PR TITLE
Upgrade ruby rake-compiler-dock images

### PR DESCRIPTION
**Pull request by apolcyn**
In particular, this picks up the fix in https://github.com/grpc/grpc/issues/28271#issuecomment-986300461

### CODE REVIEW

**Comment by apolcyn** [comment]
Ruby distrib tests are currently failing with:

```
A grpc_c.so file in this binary gRPC package is dynamically linked to libruby.
+ [[ 1 != 0 ]]
+ echo 'A grpc_c.so file in this binary gRPC package is dynamically linked to libruby.'
++ find /usr/local/rvm/gems/ruby-2.7.0 -name grpc_c.so -exec ldd '{}' ';'
++ grep -c 'not found'
+ DEPENDENCY_NOT_FOUND=1
+ [[ 1 != 0 ]]
+ echo 'A grpc_c.so file in this binary gRPC package has an non-portable dependency.'
A grpc_c.so file in this binary gRPC package has an non-portable dependency.
```

**Comment by johnnyshields** [comment]
May we get this into the 1.46.0 release?

**Comment by apolcyn** [comment]
Sorry for delay, I would like to get this change in but have not had time to solve the issue found in the test.

The problem that the test found is:

```
$ ldd grpc-1.46.0.dev-x86_64-linux/src/ruby/lib/grpc/grpc_c.so
	linux-vdso.so.1 (0x00007fffed1e6000)
	libruby.so.3.1 => not found
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc712979000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc71279f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc7131c8000)
```

So the new packages have an accidentally-built binary that's native to the build host, rather than cross-compiled. This undoes the change in https://github.com/grpc/grpc/pull/14634, which I'd prefer to avoid.

Need to figure out how to fix that, first.

**Comment by apolcyn** [comment]
Very likely we need to get https://github.com/grpc/grpc/pull/28666 in first

**Comment by apolcyn** [comment]
Temporarily cherry-picked https://github.com/grpc/grpc/pull/28666 into this branch to see if it fixes things

**Comment by apolcyn** [comment]
The latest packages still have the non-version-specific library, but it no longer has the `libruby` dependency:

```
$ find grpc-1.47.0.dev-x86_64-linux/ -name grpc_c.so
grpc-1.47.0.dev-x86_64-linux/src/ruby/lib/grpc/2.6/grpc_c.so
grpc-1.47.0.dev-x86_64-linux/src/ruby/lib/grpc/2.7/grpc_c.so
grpc-1.47.0.dev-x86_64-linux/src/ruby/lib/grpc/grpc_c.so
grpc-1.47.0.dev-x86_64-linux/src/ruby/lib/grpc/3.1/grpc_c.so
grpc-1.47.0.dev-x86_64-linux/src/ruby/lib/grpc/2.5/grpc_c.so
grpc-1.47.0.dev-x86_64-linux/src/ruby/lib/grpc/3.0/grpc_c.so
$ ldd grpc-1.47.0.dev-x86_64-linux/src/ruby/lib/grpc/grpc_c.so
	linux-vdso.so.1 (0x00007ffc59776000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f19e6e68000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f19e6c8e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f19e76c4000)
```

**Comment by jtattermusch** [comment]
+1 to building the ruby mac artifacts on linux instead of on macos (where it's slow and requires hacks). If that's what you end up doing, we should consider doing more cleanup (see my older PR https://github.com/grpc/grpc/pull/28590/files, e.g. removing the tools/distrib/build_ruby_environment_macos.sh prepare script.)

File: tools/internal_ci/macos/grpc_distribtests_ruby.sh
Comment (jtattermusch) [inline tools/internal_ci/macos/grpc_distribtests_ruby.sh:18]: +1 for keeping the job for a while as a no-op. Might come handy if we need to revert and it avoids making this PR too complex.

**Review by jtattermusch — APPROVED** [review]
LGTM as long as you've made sure that the ruby gems work well after this change is applied.
Can you summarize the test you've run and the results before merging?
This is especially important for the darwin gems, since this PR removes the possibility to build them on mac (and the rake-compiler-dock will be used from now on).

**Comment by apolcyn** [comment]
> Can you summarize the test you've run and the results before merging?
This is especially important for the darwin gems, since this PR removes the possibility to build them on mac (and the rake-compiler-dock will be used from now on).

This change shouldn't affect most packages.

I manually tested as follows:

- Running the [greeter example](https://github.com/grpc/grpc/tree/master/examples/ruby) with the linux-x86_64 packages on ruby 2.5, 3.0, and 3.1
- Ran `ldd` on all `grpc_c.so` files in the linux x86_64 package and saw there are no new or missing dependencies
- I spot checked that the ruby source-only packages are installable (on ruby 3.1).
- Running the [greeter example](https://github.com/grpc/grpc/tree/master/examples/ruby) with the `darwin-x86_64` packages on ruby 2.5, 3.0, 3.1
- Compared the output of `otool -L grpc-1.45.0-universal-darwin/src/ruby/lib/grpc/3.0/grpc_c.bundle` (for a previous macos package) with `otool -L grpc-1.47.0.dev-x86_64-darwin/src/ruby/lib/grpc/3.0/grpc_c.bundle` (a new macos package), and saw that there are no new dependencies

**Comment by apolcyn** [comment]
linux build only looks like a testing infra error:

```
kokoro_log_reader: Exhausted retries waiting for bash to return an integer exit code, returning exit code 137 instead.
```

Macos failures are due to failures of `HttpRequestTest.CancelGetDuringDNSResolution`, which is a known issue fixed by https://github.com/grpc/grpc/pull/29634

**Pull request merged**
